e2etest/l2tests/assignment.go

```diff
@@ -145,6 +148,60 @@ var _ = ginkgo.Describe("IP Assignment", func() {
 					err := cs.CoreV1().Services(svc.Namespace).Delete(context.Background(), svc.Name, metav1.DeleteOptions{})
 					return err
 				}))
+
+		ginkgo.It("should preseve the same external ip after controller restart", func() {
+			resources := config.Resources{
+				Pools: []metallbv1beta1.IPAddressPool{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "assignment-controller-reset-test-pool",
+						},
+						Spec: metallbv1beta1.IPAddressPoolSpec{
+							Addresses: []string{"192.168.10.100/32", "192.168.20.200/32"},
+						},
+					},
+				},
+			}
+			err := ConfigUpdater.Update(resources)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("creating 3 LB services")
+			jig := e2eservice.NewTestJig(cs, f.Namespace.Name, "service-specific")
+			specificService, err := jig.CreateLoadBalancerService(30*time.Second, func(svc *v1.Service) {})
+			framework.ExpectNoError(err)
+			defer service.Delete(cs, specificService)
+			service.ValidateDesiredLB(specificService)
+			allocatedIP := specificService.Status.LoadBalancer.Ingress[0].IP
+
+			svc, err := jig.CreateLoadBalancerServiceWaitForClusterIPOnly(func(svc *v1.Service) { svc.ObjectMeta.Name = "service-b" })
+			framework.ExpectNoError(err)
+			defer service.Delete(cs, svc)
+
+			svc, err = jig.CreateLoadBalancerServiceWaitForClusterIPOnly(func(svc *v1.Service) { svc.ObjectMeta.Name = "service-c" })
+			framework.ExpectNoError(err)
+			defer service.Delete(cs, svc)
+
+			ginkgo.By("restarting the controller and validating that the service keeps the same ip")
+			gomega.Consistently(func() error {
+				metallb.RestartController(cs)
+
+				gomega.Eventually(func() error {
```

Comment:
My suggestion here was to check the condition with eventually first, and then consistently, so it stabilizes.
This implementation takes for granted the result of the eventually right after the restart, not checking if it keeps being true.
Also, we are not checking the ip for the non pinned service to remain the same.

Reply:
I had the outer block as `Consistently` because I want the controller restart to be performed multiple times.
Because from my experience reproducing this issue manually, with all the moving parts here,
I had cases where only after the second controller restart the bug appeared.
Does it makes sense? Maybe we should have another inner consistently instead of the eventually?

Reply:
to validate a consistently after some time, you need to have both against the same condition.
One thing you can do is to:
- have the body in a local function (assigned to a variable)
- do the restart - eventually - consistently inside a controlled loop instead of anotehr eventually, because at the end of the day you want to try it multiple times

Reply:
good idea. I think it might be ok to remove the eventually even. and just have 5-times loop with restart and consistently asserting the services preserve their IPs.

Reply:
if you do consistently only after the restart, you might hit those "align" issue you put the sleep for though